client_integration_test.go

```diff
@@ -1489,6 +1489,49 @@ func TestClientReadFrom(t *testing.T) {
 	}
 }
 
+func TestClientReadFromSectionReader(t *testing.T) {
+	const (
+		packetSize = 1024
+		bufsize    = 4 * packetSize
+		skip       = 100
+		filesize   = bufsize - skip
+	)
+
+	sftp, cmd := testClient(t, READWRITE, NODELAY,
+		MaxPacketChecked(packetSize),
+		UseConcurrentWrites(true))
+	defer cmd.Wait()
+	defer sftp.Close()
+
+	d, err := ioutil.TempDir("", "sftptest-readfrom-sectionreader")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(d)
+
+	buf := bytes.NewReader(make([]byte, bufsize))
+	r := io.NewSectionReader(buf, 0, bufsize)
+
+	// Skip some bytes so that r.Size is no longer accurate.
+	// ReadFrom uses it to determine concurrency,
+	// but should not rely on it to determine how many bytes to send.
+	r.Seek(skip, io.SeekStart)
+
+	f := path.Join(d, "writeTest")
+	w, err := sftp.Create(f)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer w.Close()
+
+	n, err := w.ReadFrom(r)
+	assert.EqualValues(t, bufsize-skip, n)
+
+	fi, err := os.Stat(f)
+	require.NoError(t, err)
```

Comment:
Let’s not mix `if err != nil { t.Fatal("unexpected error:", err) }` and `require.NoError(t, err)` in new code. (Though, I’m sure you might be copying existing code.)

I would prefer the former rather than the later. We’ve used a lot of the later, more because we are adapting existing use of `assert.NoError(t, err)`.

Reply:
Done.